services/web/server/src/simcore_service_webserver/studies_dispatcher/_projects_permalinks.py

```diff
@@ -0,0 +1,114 @@
+import logging
+from typing import TypedDict
+
+import sqlalchemy as sa
+from aiohttp import web
+from models_library.projects import ProjectID, ProjectIDStr
+from simcore_postgres_database.models.projects import ProjectType, projects
+from simcore_service_webserver.projects.projects_exceptions import (
+    PermalinkNotAllowedError,
+    ProjectNotFoundError,
+)
+from yarl import URL
+
+from ..db import get_database_engine
+from ..projects.api import ProjectPermalink, register_permalink_factory
+from ..utils_aiohttp import create_url_for_function
+from .settings import StudiesDispatcherSettings
+
+_logger = logging.getLogger(__name__)
+
+
+# Expected inputs models
+_GroupID = str
+
+
+class _GroupAccessRightsDict(TypedDict):
+    read: bool
+    write: bool
+    delete: bool
+
+
+def create_permalink_for_study(
+    request: web.Request,
+    project_uuid: ProjectID | ProjectIDStr,
+    project_type: ProjectType,
+    project_access_rights: dict[_GroupID, _GroupAccessRightsDict],
+    project_is_public: bool,
+) -> ProjectPermalink:
+    """Checks criteria and defines how a permalink is built
+
+    Raises:
+        PermalinkNotAllowedError if some criteria are not fulfilled
+    """
+
+    # check: criterias/conditions on a project to have a permalink
+    if project_type != ProjectType.TEMPLATE:
+        raise PermalinkNotAllowedError(
+            "Can only create permalink from a template project. "
+            f"Got {project_uuid=} with {project_type=}"
+        )
+
+    if not project_access_rights.get("1", {}).get("read", False):
+        raise PermalinkNotAllowedError(
+            "Cannot create permalink if not shared with everyone. "
+            f"Got {project_uuid=} with {project_access_rights=}"
+        )
+
+    # create
+    url_for = create_url_for_function(request)
+
+    permalink: str = url_for(
+        route_name="get_redirection_to_study_page", id=f"{project_uuid}"
+    )
+
+    return ProjectPermalink(
+        url=f"{URL(permalink)}",
+        is_public=project_is_public,
+    )
+
+
+assert "uuid" in projects.columns  # nosec
+assert "type" in projects.columns  # nosec
+assert "access_rights" in projects.columns  # nosec
+assert "published" in projects.columns  # nosec
```

Comment:
I don't understand what these are for, and why here and not at the top of the module?

Reply:
If they are really required, a test might be a better place to have them.

Reply:
This is defensive programming asserts to "document" the column names which I use in permalinks implementations
These are not executed in production